File: clients/rust/src/instructions/claim_rent.rs
use crate::{
    derive_controller_authority_pda, derive_permission_pda,
    generated::instructions::ClaimRentBuilder,
};
use solana_instruction::Instruction;
use solana_pubkey::Pubkey;
use solana_sdk::system_program;

pub fn create_claim_rent_instruction(
    controller: &Pubkey,
    authority: &Pubkey,
    destination: &Pubkey,
) -> Instruction {
    let controller_authority = derive_controller_authority_pda(controller);
    let permission = derive_permission_pda(controller, authority);

    ClaimRentBuilder::new()
        .controller(*controller)
        .controller_authority(controller_authority)
        .authority(*authority)
        .permission(permission)
        .destination(*destination)
        .system_program(system_program::ID)
        .instruction()
}

File: clients/ts/src/instructions/claimRent.ts
import { Address, TransactionSigner } from "@solana/kit";
import { getClaimRentInstruction } from "../generated/instructions/claimRent";
import {
  deriveControllerAuthorityPda,
  derivePermissionPda,
} from "../pdas";
import { SVM_ALM_CONTROLLER_PROGRAM_ADDRESS } from "../generated";

/**
 * Instruction generation for claiming rent
 */
export async function createClaimRentInstruction(
  controller: Address,
  authority: TransactionSigner,
  destination: Address
) {
  const controllerAuthority = await deriveControllerAuthorityPda(controller);
  const permission = await derivePermissionPda(controller, authority.address);

  return getClaimRentInstruction(
    {
      controller,
      controllerAuthority,
      authority,
      permission,
      destination,
    },
    {
      programAddress: SVM_ALM_CONTROLLER_PROGRAM_ADDRESS,
    }
  );
}

